internal/expressions/function/functions.go

```diff
@@ -217,6 +232,70 @@ func freightMetadata(
 	}
 }
 
+func StageMetadata(
+	ctx context.Context,
+	c client.Client,
+	project string,
+) expr.Option {
+	return expr.Function(
+		"stageMetadata",
+		stageMetadata(ctx, c, project),
+		new(func(stageName, key string) any),
+	)
+}
+
+func stageMetadata(
+	ctx context.Context,
+	c client.Client,
+	project string,
+) exprFn {
+	return func(a ...any) (any, error) {
+		if len(a) != 2 {
+			return nil, fmt.Errorf("expected 2 arguments, got %d", len(a))
+		}
+
+		stageName, ok := a[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("first argument must be string, got %T", a[0])
+		}
+
+		if stageName == "" {
+			return nil, fmt.Errorf("stage name must not be empty")
+		}
+
+		key, ok := a[1].(string)
+		if !ok {
+			return nil, fmt.Errorf("second argument must be string, got %T", a[1])
+		}
+		if key == "" {
+			return nil, fmt.Errorf("metadata key must not be empty")
+		}
```

Comment:
Based on reading the implementation, I can infer usage in an expression looks something like `stageMetadata('test', 'my-info')`.

I'm wondering if UX would be nicer by taking just one argument, like so, `stageMetadata('test')`, and just returning the whole metadata map. The user can access info within the map using expressions like `stageMetadata('test')['my-info']` or `stageMetadata('test').foobar`.

Reply:
@jessesuen and I spoke offline.

@fykaa here's the direction we'll head in, please:

1. Make the second argument to `freightMetadata()` optional. When only one arg is provided, only return the map.

1. List the second arg as deprecated in the v1.8 release notes. We will remove the optional second arg in v1.10. (We typically remove deprecated things two minors later.)

1. Adjust docs to _only_ address the usage with the single argument. We don't want to promote the usage we're deprecating.

1. Implement `stageMetadata()` without the second argument from the beginning.

i.e. We'll do `stageMetadata()` the way we prefer right now and make `freightMetadata()` match it over time.

Also fwiw, @jessesuen believes the blast radius for this breaking change is very small.